### ISSUE TITLE
Inconsistent forward propagation in CPU compiled version

### ISSUE THREAD

**Issue by TomKae** [BVLC/caffe#1839]
Hello,

we are using the caffe framework on an Ubuntu 12.04 and an 14.04 system and recognized even for the same input the forward propagation function yields different output!

We use the following interface function in net.hpp:

const vector<Blob<Dtype>*>& Forward(const vector<Blob<Dtype>* > & bottom, Dtype* loss = NULL); 

which uses intern the following function of Layer.hpp:

inline Dtype Forward(const vector<Blob<Dtype>*>& bottom, vector<Blob<Dtype>*>* top); 

We already find out that this bug only appears when we compile caffe in CPU_only mode. In GPU mode, forward propagatiopn and classification, respectively, work fine.

Is this a well-known bug? Does anyone know, how to solve this problem? Is there a problem with external and operating system specific dependencies? And how can we fix it?

Would appreciate to find a solution!

Best regards, Tom

**Comment by erictzeng** [comment]
Do you mean that the forward pass in CPU mode is non-deterministic (as in, on a single OS a forward pass with the same data is producing different outputs)? Or are you comparing forward passes between different operating systems?

**Comment by TomKae** [comment]
Hi Eric,

yes, that's it! For the same input image we've got different outputs on a single OS. We started to check it in more detail and found out, that even we set "Caffe::set_phase(Caffe::TEST)" the dropout part with randomness of our network was executed for testing and thus we got inconsistent results for the same image (usually the input of the dropout layer is copied directly to the output when the TEST flag is set!). If we remove the dropout part from our net (deploy.prototxt), caffe works fine in CPU_mode, which means identical results for identical input! Furthermore, setting the TEST variable globally also works. We could not find the reason, why the phase wasn't kept "TEST" all the time! But it seems, that a singleton plays a key role. This effect happens on a Ubuntu 12.04 system, as well as on a Ubuntu 14.04 system. This error does not appear, if you compile caffe in GPU_mode. But we also recognized, that CPU_mode and GPU_mode results are slightly different for the same input. But this is another story and maybe is caused by some datatype issues.

Best, Tom

**Comment by shelhamer** [comment]
Prior to #1790 phase is a global part of Caffe state so executing between changes to phase will change the operation. #1790 makes phase an immutable part of each Net's state so that operations will be constant throughout the Net's lifetime.
